execution/engine.go

```diff
@@ -107,18 +109,56 @@ func (engine *PullEngine) Run() {
 		default:
 		}
 
-		err := engine.loop(engine.ctx, tx)
+		if engine.batchSizeManager.RecordsLeftToTake() == 0 {
```

Comment:
Add a comment here so that someone reading this doesn't have to know what `refreshTransaction` does